### PR TITLE
Delete APM fallback from Socket task operations

### PR DESCRIPTION
**Pull request by stephentoub**
The task-based methods on Socket were originally added as wrappers for the APM methods, and while a few still remain as such (namely SendTo/ReceiveMessageFromAsync), most of them are wrappers around SocketAsyncEventArgs-based implementations.  Those use a cached SAEA instance (one per operation type), but if that instance is currently in use, the operations fall back to using the APM equivalent.  This is mostly wasteful.  It's very rare to have multiple outstanding sends on the same socket concurrently, and same with receives.  It's slightly more common with accept, but even there it's not particularly valuable.

This deletes those fallback implementations and just has the operations use an SAEA; it still caches only one, and if an operation can't get a cached instance, it simply creates a new one.  Even in the rare case where multiple operations of the same type are outstanding, this is expected to be better on the whole, as while there can only be one instance cached at a time, typical use would result in a cached instance only ending up in the field for a very brief period of time, meaning one cache field could actually support a multitude of instances in use at any one moment in time.

The caching for SocketAsyncEventArgs instances was also added when the task-based APIs were new to the party, and we didn't want to negatively impact the APM/SAEA direct usage, so the task-based methods were made pay-for-play by putting them all onto a CachedEventArgs instance that was then lazily allocated.  But that makes the task-based APIs more expensive when they're actually used, and since they're now the primary use case, this PR simplifies that by moving the fields to be directly on the Socket.

cc: @geoffkizer, @scalablecory

### CODE REVIEW

**Comment by wfurt** [comment]
OSX failure is #36466

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs [hunk 1]
```diff
@@ -14,60 +13,32 @@
...
         /// <summary>Cached task with a 0 value.</summary>
         private static readonly Task<int> s_zeroTask = Task.FromResult(0);
 
-        /// <summary>Cached event args used with Task-based async operations.</summary>
-        private CachedEventArgs? _cachedTaskEventArgs;
+        /// <summary>Cached instance for accept operations.</summary>
+        private TaskSocketAsyncEventArgs<Socket>? _acceptEventArgs;
+
+        /// <summary>Cached instance for receive operations that return <see cref="ValueTask{Int32}"/>. Also used for ConnectAsync operations.</summary>
+        private AwaitableSocketAsyncEventArgs? _singleBufferReceiveEventArgs;
```
Comment (Wraith2) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:25]: Just an observation. I usually name these with the word cached in their name, `_cachedBufferReceiveEventArgs` and find that this helps distinguish working code from cache management code.

**Review by scalablecory — APPROVED** [review]
Thanks.

**Comment by geoffkizer** [comment]
Nice to see this.

**Review by geoffkizer — APPROVED** [review]

**Comment by geoffkizer** [comment]
BTW, you mentioned that SendTo/ReceiveFrom are not currently implemented using SAEA. Is there anything blocking this, to your knowledge? Or did we simply never get around to it?

**Comment by stephentoub** [comment]
> you mentioned that SendTo/ReceiveFrom are not currently implemented using SAEA. Is there anything blocking this, to your knowledge? Or did we simply never get around to it?

Nothing blocking it to my knowledge.  It's "just work" (and probably not that much), plus figuring out which SAEA should be used and how (e.g. should a new dedicated instance and field for that instance be used, should an SAEA be shared with other operations and reconfigured between each, etc.).

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:25]: Thanks.

**Pull request merged**